File: RentalModule.Application/CommandHandlers/CarOrder/CreateCarOrder/CreateCarOrderCommandHandler.cs
using RentalModule.Application.Contract.CarOrders.CreateCarOrder;
using RentalModule.Domain.Models;
using RentalModule.Domain.Models.Ids;
using RentalModule.Domain.RepositoryInterfaces.CarOrder;
using Results.Application;
using Results.Domain;
using Shared.CQRS;
using Shared.CQRS.CommandHandlers;

namespace RentalModule.Application.CommandHandlers.CarOrder.CreateCarOrder;

public class CreateCarOrderCommandHandler : ICommandHandler<CreateCarOrderCommand, HandlerResult<CreateCarOrderResponse, IErrorResult>> {
    
    private readonly ICarOrderCommandRepository _commandRepository;
    private readonly ICarOrderQueryRepository _queryRepository;

    public CreateCarOrderCommandHandler(
        ICarOrderCommandRepository commandRepository,
        ICarOrderQueryRepository queryRepository)
    {
        _commandRepository = commandRepository;
        _queryRepository = queryRepository;
    }

    public async Task<HandlerResult<CreateCarOrderResponse, IErrorResult>> Handle(CreateCarOrderCommand request,
        CancellationToken cancellationToken)
    {
        //todo check na to czy już jest jakaś oferta na tą fure
        // var carOrder =
        //     await _queryRepository.GetByIdAsync(request.CarOfferId, cancellationToken);
        // if (carOrder != null)
        //     return new EntityAlreadyExistsErrorResult()
        //     {
        //         Title = "Cannot create car order",
        //         Message =
        //             $"Car order for the car offer {request.CarOfferId} and user {request.UserId} already exists in the database."
        //     };

        var newCarOrderId = new CarOrderId(Guid.NewGuid());
        var carOrderToAdd = new CarOrderModel
        {
            Id = newCarOrderId,
            UserId = request.UserId,
            CarOfferId = request.CarOfferId,
            StartDate = request.StartDate,
            EndDate = request.EndDate,
            Notes = request.Notes ,
            NumOfDrivers = request.NumOfDrivers,
            TotalCost = request.TotalCost
        };

        var addedCarOrder = await _commandRepository.AddAsync(carOrderToAdd, cancellationToken);

        return new CreateCarOrderResponse()
        {
            Id = addedCarOrder.Id,
            Title = "Car order created",
            Message = $"Car order for '{addedCarOrder.CarOfferId}' was created in the database."
        };
    }
}

File: RentalApi/Application/command_handlers/car_order/create_car_order/create_car_order_command_handler.go
package commands

import (
	"context"
	contract "rental-api/Application.contract/car_orders/CreateCarOrder"
	models "rental-api/Domain/models/domestic"
	car_offer_repository "rental-api/Domain/repository_interfaces/car_offer_repository"
	repository_interfaces "rental-api/Domain/repository_interfaces/car_order_repository"
	"rental-api/Domain/responses"
	data_fetcher "rental-api/Domain/service_interfaces"
)

type CreateCarOrderCommandHandler struct {
	carOrderCommandRepository repository_interfaces.CarOrderCommandRepository
	carOfferQueryRepository   car_offer_repository.CarOfferQueryRepository
	dataFetcher               data_fetcher.MicroserviceConnector
}

func NewCreateCarOrderCommandHandler(
	carOrderCommandRepository repository_interfaces.CarOrderCommandRepository,
	carOfferQueryRepository car_offer_repository.CarOfferQueryRepository,
	dataFetcher data_fetcher.MicroserviceConnector,
) *CreateCarOrderCommandHandler {
	return &CreateCarOrderCommandHandler{
		carOrderCommandRepository: carOrderCommandRepository,
		carOfferQueryRepository:   carOfferQueryRepository,
		dataFetcher:               dataFetcher,
	}
}

func (h *CreateCarOrderCommandHandler) Handle(ctx context.Context, command *CreateCarOrderCommand) (*contract.CreateCarOrderResponse, error) {
	userInfo, err := h.dataFetcher.GetUserInternalInfo(command.JwtToken)
	if err != nil {
		response := responses.NewResponse[contract.CreateCarOrderResponse](401, "Unauthorized")
		return &response, nil
	}

	carOffer, err := h.carOfferQueryRepository.GetCarOfferByID(command.CarOfferId)
	if err != nil || carOffer == nil {
		response := responses.NewResponse[contract.CreateCarOrderResponse](404, "Car offer not found")
		return &response, nil
	}

	carOrder := &models.CarOrderModel{
		UserId:           userInfo.ID,
		CarOfferId:       command.CarOfferId,
		StartDate:        command.StartDate,
		EndDate:          command.EndDate,
		DeliveryLocation: command.DeliveryLocation,
		ReturnLocation:   command.ReturnLocation,
		NumOfDrivers:     command.NumOfDrivers,
		TotalCost:        command.TotalCost,
	}

	result, err := h.carOrderCommandRepository.CreateCarOrder(ctx, carOrder)
	if err != nil {
		response := responses.NewResponse[contract.CreateCarOrderResponse](500, "Failed to create car order")
		return &response, nil
	}

	return &contract.CreateCarOrderResponse{
		BaseResponse: responses.NewBaseResponse(200, "Car order created successfully"),
		Id:           *result,
	}, nil
}
